libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs

```diff
@@ -0,0 +1,219 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Test.Cardano.Ledger.STS where
+
+import Control.State.Transition.Extended
+import Data.Default.Class (Default (def))
+import Data.Foldable
+import Lens.Micro ((^.))
+
+import Test.QuickCheck
+
+import Test.Tasty
+import Test.Tasty.QuickCheck
+
+import Cardano.Ledger.Conway
+import Cardano.Ledger.Conway.Core
+import Cardano.Ledger.Conway.Rules
+
+import Cardano.Ledger.Conway.PParams
+
+import Test.Cardano.Ledger.Generic.Proof
+import Test.Cardano.Ledger.Shelley.Utils
+
+import Cardano.Ledger.CertState
+import Test.Cardano.Ledger.Constrained.Ast
+import Test.Cardano.Ledger.Constrained.Preds.CertState
+import Test.Cardano.Ledger.Constrained.Preds.PParams
+import Test.Cardano.Ledger.Constrained.Preds.Universes
+import Test.Cardano.Ledger.Constrained.Rewrite
+import Test.Cardano.Ledger.Constrained.TypeRep
+import Test.Cardano.Ledger.Constrained.Vars
+
+import qualified Data.Map as Map
+import Data.Maybe
+
+import Test.Cardano.Ledger.Constrained.Env
+
+import Test.Cardano.Ledger.Constrained.Classes
+import Test.Cardano.Ledger.Constrained.Monad
+
+import Test.Cardano.Ledger.Constrained.Shrink
+
+import Test.Cardano.Ledger.Alonzo.AlonzoEraGen ()
+import Test.Cardano.Ledger.Alonzo.EraMapping ()
+import Test.Cardano.Ledger.Constrained.Solver
+
+import Cardano.Ledger.Conway.TxCert
+
+------------------------------------------------------------------------
+-- Generator Setup
+------------------------------------------------------------------------
+
+genFromConstraints :: Era era => Proof era -> OrderInfo -> [Pred era] -> RootTarget era s t -> Gen t
+genFromConstraints proof order cs target = do
+  subst <- genSubstFromConstraints proof order cs
+  env <- monadTyped $ substToEnv subst emptyEnv
+  monadTyped (runTarget env target)
+
+genSubstFromConstraints :: Era era => Proof era -> OrderInfo -> [Pred era] -> Gen (Subst era)
+genSubstFromConstraints proof order cs = do
+  (_, graph) <- compileGenWithSubst order emptySubst cs
+  result <- genDependGraph False proof graph
+  case result of
+    Left errs -> error $ unlines errs
+    Right subst -> pure subst
+
+shrinkFromConstraints :: Era era => OrderInfo -> [Pred era] -> RootTarget era t t -> t -> [t]
+shrinkFromConstraints order cs target val = do
+  let env = saturateEnv (getTarget val target emptyEnv) cs
+  graph <- monadTyped $ compile order $ rewrite cs
+  env' <- shrinkEnv graph env
+  monadTyped $ runTarget env' target
```

Comment:
Be nice if we had some comments here, something like
  let env = saturateEnv (getTarget val target emptyEnv) cs
  -- Build a new env, with just the variables in the target, where the vars are bound to the values extracted from 'val'
 graph <- monadTyped $ compile order $ rewrite cs
-- Build a new graph from the original constraints
  env' <- shrinkEnv graph env
-- Shrink each of the values bound in the env, by using the graph to determine what other variables they depended on
 monadTyped $ runTarget env' target
-- Rebuild the 'val' using the target and the new env with the shrunk  values.